cirq/circuits/circuit_operation.py

```diff
@@ -0,0 +1,382 @@
+                for minimizing the total number of operations in a circuit.
+            validate_modulus: For use with `modulus`. If this is True, the
+                function will raise an error if the provided modulus does not
+                match the actual modulus of this operation.
+
+        Returns:
+            A copy of this operation repeated 'repetitions' times.
+
+        Raises:
+            TypeError: `repetitions` is not an integer value.
+            NotImplementedError: The operation contains measurements and
+                cannot have repetitions.
+            ValueError: `validate_modulus` was set to True, but raising this
+                operation to `modulus` does not produce the identity.
+        """
+        if not isinstance(repetitions, (int, np.integer)):
+            raise TypeError('Only integer repetitions are allowed.')
+        repetitions = int(repetitions)
+        if protocols.is_measurement(self.circuit):
+            raise NotImplementedError('Loops over measurements are not supported.')
+        new_op = self.copy()
+        new_op._repetitions *= repetitions
+        if modulus:
+            if validate_modulus and not linalg.allclose_up_to_global_phase(
+                protocols.unitary(self.circuit * modulus), np.eye(np.prod(self.circuit.qid_shape()))
+            ):
+                raise ValueError('Raising the circuit to "modulus" must produce the identity.')
+            new_op._repetitions %= modulus
+            if allow_invert and new_op._repetitions > modulus // 2:
+                new_op._repetitions -= modulus
+        return new_op
+
+    def __pow__(self, power: int) -> 'CircuitOperation':
+        return self.repeat(power)
+
+    def with_qubit_mapping(self, qubit_map: Dict['cirq.Qid', 'cirq.Qid']) -> 'CircuitOperation':
+        """Returns a copy of this operation with an updated qubit mapping.
+
+        Args:
+            qubit_map: A mapping of old qubits to new qubits. This map will be
+                composed with any existing qubit mapping.
+
+        Returns:
+            A copy of this operation targeting qubits as indicated by qubit_map.
+        """
+        new_op = self.copy()
+        new_op._qubit_map = {k: qubit_map.get(v, v) for k, v in new_op.qubit_map.items()}
+        return new_op
+
+    def with_qubits(self, *new_qubits: 'cirq.Qid'):
+        """Returns a copy of this operation with an updated qubit mapping.
+
+        Args:
+            new_qubits: A list of qubits to target. Qubits in this list are
+                matched to qubits in the circuit following default qubit order,
+                ignoring any existing qubit map.
+
+        Returns:
+            A copy of this operation targeting `new_qubits`.
+
+        Raises:
+            ValueError: `new_qubits` has a different number of qubits than
+                this operation.
+        """
+        if len(new_qubits) != protocols.num_qubits(self.circuit):
```

Comment:
nit: you're running `num_qubits` twice on the error path... we could save that repetition by pulling `expected` out in front of the if